src/parameterizations/lateral/MOM_hor_visc.F90

```diff
@@ -828,133 +838,247 @@ subroutine horizontal_viscosity(u, v, h, diffu, diffv, MEKE, VarMix, G, GV, US,
 
     meke_res_fn = 1.
 
-    do j=Jsq,Jeq+1 ; do i=Isq,Ieq+1
-      if ((CS%Smagorinsky_Kh) .or. (CS%Smagorinsky_Ah)) then
-        Shear_mag = sqrt(sh_xx(i,j)*sh_xx(i,j) + &
-          0.25*((sh_xy(I-1,J-1)*sh_xy(I-1,J-1) + sh_xy(I,J)*sh_xy(I,J)) + &
-                (sh_xy(I-1,J)*sh_xy(I-1,J) + sh_xy(I,J-1)*sh_xy(I,J-1))))
+    if ((CS%Smagorinsky_Kh) .or. (CS%Smagorinsky_Ah)) then
+      do j=Jsq,Jeq+1 ; do i=Isq,Ieq+1
+        sh_xx_sq = sh_xx(i,j) * sh_xx(i,j)
+        sh_xy_sq = 0.25 * ( &
+            (sh_xy(I-1,J-1) * sh_xy(I-1,J-1) + sh_xy(I,J) * sh_xy(I,J)) &
+            + (sh_xy(I-1,J) * sh_xy(I-1,J) + sh_xy(I,J-1) * sh_xy(I,J-1)) &
+        )
+        Shear_mag(i,j) = sqrt(sh_xx_sq + sh_xy_sq)
+      enddo ; enddo
+    endif
+
+    if (CS%better_bound_Ah .or. CS%better_bound_Kh) then
+      do j=Jsq,Jeq+1 ; do i=Isq,Ieq+1
+        h_min = min(h_u(I,j), h_u(I-1,j), h_v(i,J), h_v(i,J-1))
+        hrat_min(i,j) = min(1.0, h_min / (h(i,j,k) + h_neglect))
+      enddo ; enddo
+
+      if (CS%better_bound_Kh) then
+        do j=Jsq,Jeq+1 ; do i=Isq,Ieq+1
+          visc_bound_rem(i,j) = 1.0
+        enddo ; enddo
       endif
+    endif
+
+    if (CS%Laplacian) then
       if ((CS%Leith_Kh) .or. (CS%Leith_Ah)) then
         if (CS%use_QG_Leith_visc) then
-          vert_vort_mag = MIN(grad_vort_mag_h(i,j) + grad_div_mag_h(i,j),3.*grad_vort_mag_h_2d(i,j))
+          do j=Jsq,Jeq+1 ; do i=Isq,Ieq+1
+            grad_vort = grad_vort_mag_h(i,j) + grad_div_mag_h(i,j)
+            grad_vort_qg = 3. * grad_vort_mag_h_2d(i,j)
+            vert_vort_mag(i,j) = min(grad_vort, grad_vort_qg)
+          enddo ; enddo
         else
-          vert_vort_mag = (grad_vort_mag_h(i,j) + grad_div_mag_h(i,j))
+          do j=Jsq,Jeq+1 ; do i=Isq,Ieq+1
+            vert_vort_mag(i,j) = grad_vort_mag_h(i,j) + grad_div_mag_h(i,j)
+          enddo ; enddo
         endif
       endif
-      if (CS%better_bound_Ah .or. CS%better_bound_Kh) then
-        hrat_min = min(1.0, min(h_u(I,j), h_u(I-1,j), h_v(i,J), h_v(i,J-1)) / &
-                            (h(i,j,k) + h_neglect) )
-        visc_bound_rem = 1.0
-      endif
 
-      if (CS%Laplacian) then
-        ! Determine the Laplacian viscosity at h points, using the
-        ! largest value from several parameterizations.
-        Kh = CS%Kh_bg_xx(i,j) ! Static (pre-computed) background viscosity
+      ! Determine the Laplacian viscosity at h points, using the
+      ! largest value from several parameterizations.
+
+      ! Static (pre-computed) background viscosity
+      do j=Jsq,Jeq+1 ; do i=Isq,Ieq+1
+        Kh(i,j) = CS%Kh_bg_xx(i,j)
+      enddo ; enddo
+
+      do j=Jsq,Jeq+1 ; do i=Isq,Ieq+1
         if (CS%add_LES_viscosity) then
-          if (CS%Smagorinsky_Kh) Kh = Kh + CS%Laplac2_const_xx(i,j) * Shear_mag
-          if (CS%Leith_Kh) Kh = Kh + CS%Laplac3_const_xx(i,j) * vert_vort_mag*inv_PI3
+          if (CS%Smagorinsky_Kh) &
```

Comment:
Would the code in the 13-line block from the new lines 888 - 900 be a good candidate for restructuring to put the logical tests outside of the do-loops, as was found to be more efficient elsewhere in this PR?

Reply:
Yes, this is one example of several others, but unfortunately I started to experience some of the "stack spill" issues that we discussed privately.

At this point, I am too worried to make further changes until we figure out how to reduce the function stack (and perhaps confirm a little more thoroughly that this is what is actually going on, and that reducing the function stack would help here)

Reply:
This comment block appears later in a similar section.  Perhaps another disclaimer is needed here...?
```
! NOTE: The following do-block can be decomposed and vectorized, but
!   appears to cause slowdown on some machines.  Evidence suggests that
!   this is caused by excessive spilling of stack variables.
! TODO: Vectorize these loops after stack usage has been reduced..
```